payments/utils.py

```diff
@@ -0,0 +1,147 @@
+from decimal import Decimal, ROUND_HALF_UP
+from functools import wraps
+from typing import Callable
+
+from dateutil.relativedelta import relativedelta
+from dateutil.rrule import MONTHLY, rrule
+
+from leases.utils import (
+    calculate_berth_lease_end_date,
+    calculate_berth_lease_start_date,
+)
+
+
+def round_price(price):
+    return price.quantize(Decimal("0.01"), rounding=ROUND_HALF_UP)
+
+
+def rounded(func):
+    """
+    Decorator for rounding function result
+    """
+
+    @wraps(func)
+    def wrapped(*args, **kwargs):
+        value = func(*args, **kwargs)
+        return round_price(value)
+
+    return wrapped
+
+
+def currency_format(value):
+    return f"{round_price(value)}€"
+
+
+def currency(func):
+    @wraps(func)
+    def wrapped(*args, **kwargs):
+        value = func(*args, **kwargs)
+        return currency_format(value)
+
+    return wrapped
+
+
+def percentage_format(value):
+    return f"{round_price(value)}%"
+
+
+def percentage(func):
+    @wraps(func)
+    def wrapped(*args, **kwargs):
+        value = func(*args, **kwargs)
+        return percentage_format(value)
+
+    return wrapped
+
+
+def convert_aftertax_to_pretax(aftertax_price, tax_percentage):
+    return aftertax_price / (1 + tax_percentage / 100)
+
+
+def calculate_order_due_date(delta=14):
+    return date.today() + timedelta(days=delta)
+
+
+def _calculate_price_for_partial_month(
+    month_date: date, base_price: Decimal, days: Callable[[int], int]
+) -> Decimal:
+    _, days_in_month = calendar.monthrange(month_date.year, month_date.month)
+    price_per_day = base_price / days_in_month
+    return days(days_in_month) * price_per_day
+
+
+@rounded
+def calculate_product_partial_month_price(
+    base_price: Decimal, start_date: date, end_date: date
+) -> Decimal:
+    price = Decimal(0)
+
+    # If the dates are on the same month/year
+    if start_date.month == end_date.month and start_date.year == end_date.year:
+        price = _calculate_price_for_partial_month(
+            start_date, base_price, days=lambda _: (end_date - start_date).days
+        )
+    else:
+        # Calculate price for the start month
+        price += _calculate_price_for_partial_month(
+            start_date,
+            base_price,
+            days=lambda days_in_month: abs(start_date.day - days_in_month),
+        )
+
+        # Calculate price for the months in between
+        months_between = list(
+            rrule(
+                MONTHLY,
+                dtstart=start_date + relativedelta(day=1, months=1),
+                until=end_date - relativedelta(day=1, months=1),
+            )
+        )
+        # Calculate the added price for all the complete months in between
+        price += len(months_between) * base_price
+
+        # Calculate price for the end month
+        price += _calculate_price_for_partial_month(
+            end_date,
+            base_price,
+            days=lambda days_in_month: abs(days_in_month - start_date.day),
```

Comment:
and here I think you just need to pass the `lambda _: end_date.day`

Reply:
hm, I kinda feel like here it should be then `lambda _: end_date.day - 1` since prooobably excluding the last day, but maybe we should check with Ville